lib/discordgo/permission.go

```diff
@@ -1,53 +1,99 @@
-	PermissionChangeNickname          int64 = 0x00004000000                      // (1 << 26)	Allows for modification of own nickname
-	PermissionManageNicknames         int64 = 0x00008000000                      // (1 << 27)	Allows for modification of other users nicknames
-	PermissionManageRoles             int64 = 0x00010000000                      // (1 << 28)	Allows management and editing of roles	T, V, S
-	PermissionManageWebhooks          int64 = 0x00020000000                      // (1 << 29)	Allows management and editing of webhooks	T
-	PermissionManageEmojisAndStickers int64 = 0x00040000000                      // (1 << 30)	Allows management and editing of emojis and stickers
-	PermissionManageEmojis            int64 = PermissionManageEmojisAndStickers  // deprecated: use PermissionManageEmojisAndStickers
-	PermissionUseSlashCommands        int64 = PermissionUseApplicationCommands   // deprecated: use PermissionUseApplicationCommands
-	PermissionUseApplicationCommands  int64 = 0x00080000000                      // (1 << 31)	Allows members to use slash commands in text channels	T
-	PermissionRequestToSpeak          int64 = 0x00100000000                      // (1 << 32)	Allows for requesting to speak in stage channels. (This permission is under active development and may be changed or removed.)	S
-	PermissionManageEvents            int64 = 0x00200000000                      // (1 << 33)   Allows for creating, editing, and deleting scheduled events	V, S
-	PermissionManageThreads           int64 = 0x00400000000                      // (1 << 34)	Allows for deleting and archiving threads, and viewing all private threads	T
-	PermissionUsePublicThreads        int64 = 0x00800000000                      // (1 << 35)	Allows for creating and participating in threads	T
-	PermissionUsePrivateThreads       int64 = 0x01000000000                      // (1 << 36)	Allows for creating and participating in private threads	T
-	PermissionUseExternalStickers     int64 = 0x02000000000                      // (1 << 37)	Allows the usage of custom stickers from other servers	T
-	PermissionSendMessagesInThreads   int64 = 0x04000000000                      // (1 << 38)   Allows for sending messages in threads	T
-	PermissionUseEmbeddedActivities   int64 = 0x08000000000                      // (1 << 39)   Allows for using Activities (applications with the EMBEDDED flag) in a voice channel	V
-	PermissionModerateMembers         int64 = 0x10000000000                      // (1 << 40)   Allows for timing out users to prevent them from sending or reacting to messages in chat and threads, and from speaking in voice and stage channels
-
-	// all bits set except the leftmost to avoid using negative numbers in case discord dosen't handle it
-	PermissionAll int64 = int64(^uint64(0) >> 1)
+	PermissionCreateInstantInvite     int64 = 0x00000000001 // (1 << 0)		Allows creation of instant invites	T, V, S
```

Comment:
just interested, why not use 1 << 0 instead, and not 0x....001 and so on

Reply:
I don't know why the existing code uses hexadecimal either. My best guess is that it is copied from the [permission flags table in the Discord API documentation](https://discord.com/developers/docs/topics/permissions), but that also provides the left-shift expressions so ¯\\\_(ツ)\_/¯. I personally would have written this using right shifts, but seeing that the diff for this PR is large enough already I figured it'd be nice to not switch too many things up.

Reply:
maybe add the ones after 1 << 40 as well from official docs, for completeness sake

Reply:
Good point, thanks. I originally thought that the only ones missing were the new poll-related permissions so was planning on adding those in another PR for poll support in general, but seeing we're missing quite a few I will go ahead and add those tomorrow.

Reply:
I added the new permissions in https://github.com/botlabs-gg/yagpdb/pull/1649/commits/d1a828d8d370ff401cc75d151a21a2c733a2d1fb, which is also a nice example of the benefits of this refactor (per commit description.)

While adding these new permissions, I thought about the issue re: hexadecimal you brought up earlier some more. Originally, like I said, I planned on keeping the code format untouched, but adding the new values in hexadecimal was rather inconvenient since Discord API's doc quotes them to 15 digits: `0x0002000000000000`, for instance. This messed up the alignment and made lines stretch to ridiculous levels after `gofmt`'ing, so I decided to switch everything to use left-shifts in https://github.com/botlabs-gg/yagpdb/pull/1649/commits/4b38bf1cc6cc849af59d3dc13c3a6033e25b85df.

Related: I did some digging in the commit history of `bwmarrin/discordgo` to figure out why they are using hexadecimal values, and found https://github.com/bwmarrin/discordgo/commit/843c765ae3eeecaf8e9b808ea8e3da10ba600f91. They say
> We define our permissions constants based on Discord's values,
and thus the use of `iota` is fragile, makes docs harder to read,
and makes changes touching these constants harder to review.
Here, we now use the full 64-bit representation of each constant,
which both removes the implicit ordering dependencies and makes
the values easier to immediately see.

I agree with the point re: no `iota` -- it makes the code very fragile and is prone to cause issues when changed -- for instance, `PermissionSendPolls = 1 << 49` is right after `PermissionSendVoiceMessages = 1 << 46` and using `iota` carelessly would erroneously set it to `1 << 47`. However, I think that using the explicit left-shift representation still provides the benefits described (easy to review and to immediately see values.)

Reply:
I checked that I did not make any copy-paste errors while switching to the left-shift representation using the following script: https://go.dev/play/p/zlzrZOZquF7. The permission constants are copied from the new file and the series of Println checks are generated by taking the old file and substituting `s/(\w+)\s+int64 = (0x\d+)/fmt.Println("$1", $1 == $2)/g`.

Reply:
looks good and works on my testing, only minor issue I have are file names > permissions.go (like dstate has) and permission_names.go, plural just feel better to me, but this is non-issue